sqlccl: add Load function

madelynnblue opened the pull request:
Load converts SQL files into SST files of the form expected by
Restore. Currently not hooked up to any user-facing feature, but
used by some benchmarks to speed up database initialization before
benchmarking Restore.

go test  -tags '' -run "-" -bench "Import" -timeout 5m ./pkg/ccl/sqlccl -benchtime 20s
BenchmarkImport-4          	 5000000	      6885 ns/op	  16.85 MB/s
BenchmarkImportRestore-4   	 1000000	     31695 ns/op	   3.63 MB/s
BenchmarkImportSQL-4       	 1000000	     37827 ns/op	   3.04 MB/s

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13495)
<!-- Reviewable:end -->

danhhz commented:
Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/load.go, line 41 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13495#-KcUiDEVPiM0ASRCUF2z:-KcUiDEVPiM0ASRCUF3-:brwikg5) ([raw file](https://github.com/cockroachdb/cockroach/blob/db30ceb788f02038d448392862f99c8073bfe7b4/pkg/ccl/sqlccl/load.go#L41)):*
> ```Go
> 	ctx context.Context, db *gosql.DB, r io.Reader, database, uri string, ts hlc.Timestamp,
> ) (BackupDescriptor, error) {
> 	importChunkBytes := config.DefaultZoneConfig().RangeMaxBytes / 2
> ```

This needs to be a param so TestImportChunking can set it smaller. That test takes FOREVER

---

*[pkg/ccl/sqlccl/load.go, line 48 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13495#-KcUiQe_nH1uVA25SqqK:-KcUiQeajFup4OG85Nmy:b-gpq3hr) ([raw file](https://github.com/cockroachdb/cockroach/blob/db30ceb788f02038d448392862f99c8073bfe7b4/pkg/ccl/sqlccl/load.go#L48)):*
> ```Go
> 	evalCtx := parser.EvalContext{}
> 
> 	base, err := storageccl.ExportStorageFromURI(ctx, uri)
> ```

we call this "dir" instead of "base" everywhere else

---

*[pkg/ccl/sqlccl/load.go, line 63 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13495#-KcUice6rNZvrZcHeMgw:-KcUice6rNZvrZcHeMgx:bdvu78j) ([raw file](https://github.com/cockroachdb/cockroach/blob/db30ceb788f02038d448392862f99c8073bfe7b4/pkg/ccl/sqlccl/load.go#L63)):*
> ```Go
> 			d.descriptor
> 		FROM system.namespace n INNER JOIN system.descriptor d ON n.id = d.id
> 		WHERE n.parentID = 0
> ```

use keys.RootNamespaceID instead of 0 here

---

*[pkg/ccl/sqlccl/load.go, line 162 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13495#-KcUkIB35CDzi3ChQ7YO:-KcUkIB35CDzi3ChQ7YP:bdu1cat) ([raw file](https://github.com/cockroachdb/cockroach/blob/db30ceb788f02038d448392862f99c8073bfe7b4/pkg/ccl/sqlccl/load.go#L162)):*
> ```Go
> 				return BackupDescriptor{}, errors.Errorf("expected previous CREATE TABLE %s statement", name)
> 			}
> 			if name != tableName {
> ```

ReNormalizeName on both sides of this

---

*[pkg/ccl/sqlccl/load.go, line 168 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13495#-KcUkfE5fsJ-Ecay5KRs:-KcUk_ta290vdJyVXXsT:b-mseznx) ([raw file](https://github.com/cockroachdb/cockroach/blob/db30ceb788f02038d448392862f99c8073bfe7b4/pkg/ccl/sqlccl/load.go#L168)):*
> ```Go
> 			err := insertStmtToKVs(ctx, tableDesc, defaultExprs, cols, evalCtx, ri, s, func(kv roachpb.KeyValue) {
> 				if prevKey.Compare(kv.Key) >= 0 {
> 					outOfOrder = true
> ```

return early (and probably do `outOfOrder || prevKey.Compare(kv.Key) >= 0` in the if condition

---

*[pkg/ccl/sqlccl/load.go, line 233 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13495#-KcUkxpay3llzRpDtCot:-KcUkxpay3llzRpDtCou:b-6rezbf) ([raw file](https://github.com/cockroachdb/cockroach/blob/db30ceb788f02038d448392862f99c8073bfe7b4/pkg/ccl/sqlccl/load.go#L233)):*
> ```Go
> ) error {
> 	if stmt.OnConflict != nil {
> 		return errors.Errorf("import insert: ON CONFLICT not supported: %q", stmt)
> ```

s/import/load/ here and elsewhere

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13495)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Oh and maybe rewrite BenchmarkClusterBackup with the version in https://github.com/cockroachlabs/enterprise/blob/master/pkg/ccl/sqlccl/backup_test.go

mjibson commented:
Fixed BenchmarkClusterBackup too.

---

Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/ccl/sqlccl/load.go, line 41 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13495#-KcUiDEVPiM0ASRCUF2z:-KcZB4G5I-Hff50Ac1sV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db30ceb788f02038d448392862f99c8073bfe7b4/pkg/ccl/sqlccl/load.go#L41)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

This needs to be a param so TestImportChunking can set it smaller. That test takes FOREVER
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/load.go, line 48 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13495#-KcUiQe_nH1uVA25SqqK:-KcZBKG46fCM34vHVfIx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db30ceb788f02038d448392862f99c8073bfe7b4/pkg/ccl/sqlccl/load.go#L48)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

we call this "dir" instead of "base" everywhere else
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/load.go, line 63 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13495#-KcUice6rNZvrZcHeMgw:-KcZBpDEQDA8AeiwpdSY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db30ceb788f02038d448392862f99c8073bfe7b4/pkg/ccl/sqlccl/load.go#L63)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

use keys.RootNamespaceID instead of 0 here
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/load.go, line 162 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13495#-KcUkIB35CDzi3ChQ7YO:-KcZCRiWKFFhfLf8DFUl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db30ceb788f02038d448392862f99c8073bfe7b4/pkg/ccl/sqlccl/load.go#L162)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

ReNormalizeName on both sides of this
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/load.go, line 168 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13495#-KcUkfE5fsJ-Ecay5KRs:-KcZCnPfu2PiY_E3e-iB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db30ceb788f02038d448392862f99c8073bfe7b4/pkg/ccl/sqlccl/load.go#L168)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

return early (and probably do `outOfOrder || prevKey.Compare(kv.Key) >= 0` in the if condition
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/load.go, line 233 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13495#-KcUkxpay3llzRpDtCot:-KcZC1oH9jQPiYeCflc9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db30ceb788f02038d448392862f99c8073bfe7b4/pkg/ccl/sqlccl/load.go#L233)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

s/import/load/ here and elsewhere
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13495#-:-KcZI0me-zDRqC6D9pfu:b-8h5046)*
<!-- Sent from Reviewable.io -->

danhhz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I think you'll need to `pull --rebase` to get depcheck to pass

---

Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/ccl/sqlccl/load_test.go, line 36 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13495#-KcZKjkYJmrpaI0OafVr:-KcZJNBqdFWBQr7QyNK0:bxesu5e) ([raw file](https://github.com/cockroachdb/cockroach/blob/e91e4f2d573d3016bbeab86b07936d21e55de563/pkg/ccl/sqlccl/load_test.go#L36)):*
> ```Go
> func TestImportChunking(t *testing.T) {
> 	// Generate at least 2 chunks.
> 	const chunkSize = 1024 * 500 // 100KB
> ```

Isn't this 500KB? You can probably just delete the comment

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13495#-:-KcZKlZGgdM11PiabgGd:b-14srwg)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sqlccl: add Load function

Load converts SQL files into SST files of the form expected by
Restore. Currently not hooked up to any user-facing feature, but
used by some benchmarks to speed up database initialization before
benchmarking Restore.

go test  -tags '' -run "-" -bench "Import" -timeout 5m ./pkg/ccl/sqlccl -benchtime 20s
BenchmarkImport-4          	 5000000	      6885 ns/op	  16.85 MB/s
BenchmarkImportRestore-4   	 1000000	     31695 ns/op	   3.63 MB/s
BenchmarkImportSQL-4       	 1000000	     37827 ns/op	   3.04 MB/s

go test  -tags '' -run "-" -bench "BenchmarkClusterBackup" -timeout 5m ./pkg/ccl/sqlccl -benchtime 1s -v
BenchmarkClusterBackup-4   	   50000	     62479 ns/op	   1.84 MB/s

mjibson commented:
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/ccl/sqlccl/load_test.go, line 36 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13495#-KcZKjkYJmrpaI0OafVr:-Kc_-PvD4MsOAtzWrymL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e91e4f2d573d3016bbeab86b07936d21e55de563/pkg/ccl/sqlccl/load_test.go#L36)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Isn't this 500KB? You can probably just delete the comment
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13495)*
<!-- Sent from Reviewable.io -->

mjibson commented:
@RaduBerinde we're merging this because we need it for some benchmark work (it's not exposed to users yet). But we hope that you're able to review it at some point anyway.

RaduBerinde commented on pkg/ccl/sqlccl/load.go:
```diff
@@ -0,0 +1,368 @@
+	evalCtx parser.EvalContext,
+	ri sql.RowInserter,
+	stmt *parser.Insert,
+	f func(roachpb.KeyValue),
```
Could use some documentation as to what `f` is. Actually, it would make things easier if we defined this inside `insertStmtToKVs` and instead pass and return `kvs`.

RaduBerinde reviewed (commented):
LGTM! Pretty cool stuff. I only have a minor comment, feel free to ignore it.
